Add method RDSE.check_parameters()

ctrl-z-9000-times opened the pull request:
Hi everyone,

This PR adds a new check to the Random Distributed Scalar Encoder for sane Parameters. Bad RDSE parameters cause the RDSE to perform poorly and even to totally fail. The new checks are called automatically whenever an RDSE is created.

This works by experimentally measuring the hash collision probability. Currently it allows up to 33% of the encodings to collide before it returns an error, and it checks 1000 different random encodings for collisions.

For speed, this method is skipped if the RDSE parameters are obviously good. The heuristics for skipping the check are:
* size >= 1000 and
* activeBits >= 10

Fixes #777

dkeeney reviewed (commented):
After studying this I see why this would be a good thing to have.  This might help with the rare random failures some folks have seen.

You said that this new functions should be called in the constructor. My first impulse was that this would be a large overhead but actually the RDSE would not be created that often.  So, yes.  Lets add this to the constructor, perhaps at the end of the initialize() method.  I do not see that you have done that.

ctrl-z-9000-times commented:
Hi dkeeney, I updated the initial commit (force push) to make it call the check during the constructor.

I also added a heuristic to avoid the hash collision check if the RDSE is obviously large enough.
This should make it run fast BC the run time of the check is directly related to the size of the RDSE and now it skips the check for very large RDSE's.

dkeeney reviewed (approved):
Looks good.   Thanks for adding this.

dkeeney commented:
I don't think the build process is running.
I do not have admin permissions (as far as I know) so I cannot look into this.
@breznak can you check this?

breznak commented on src/htm/encoders/RandomDistributedScalarEncoder.cpp:
```diff
@@ -118,14 +120,30 @@ void RandomDistributedScalarEncoder::encode(Real64 input, SDR &output)
     // they're handled.
 
-    // Exercise for the reader: Calculate the probability of a hash collision
-    // and account for it in the sparsity.
```
never liked this! :+1:  
so..were the collisions more common than we expected (in settings people actually used)?

ctrl-z-9000-times replied:
HTM systems require a lot of cells in order to work properly.

I think there is a tendency for people (myself included) to make small HTM systems, either because
- it runs faster, 
- or they think it will be easier to understand/debug (in this case you're better off using a regular scalar encoder).

And even though I understand all this, I'm still tempted to reduce the number of cells in my experiments because I'm impatient and want it to run faster.

breznak commented on src/htm/encoders/RandomDistributedScalarEncoder.cpp:
```diff
@@ -118,14 +120,30 @@ void RandomDistributedScalarEncoder::encode(Real64 input, SDR &output)
+    return true;
+  }
+  const auto maximum_overlap = 0.33f;
+  const auto number_of_trials = 1000u;
```
nit: this could have been default params to the constructor, or just set in init as `self.max_overlap`,..

ctrl-z-9000-times replied:
Hi Breznak,

Yes but I disagree for two reasons:

1) I tuned the heuristic to skip this method manually by trial and error. If the user changes the `maximum_overlap` then this heuristic might not be valid anymore (it would allow the parameters even though there is a chance that they are bad).

2) I don't think it would be useful for the users. My goal here is just to sanity check the users input for obviously bad parameters. I think that if the users want to control the `maximum_overlap` between two RDSE representations that they would be better off spending some time playing around with SDR's in order to gain an intuitive understanding how SDR size & sparsity affect the probability of collisions between random SDRs.

breznak reviewed (dismissed):
Thanks David! Long time no see, glad you came with the fix for RDSE. 
Approved,

> I do not have admin permissions (as far as I know) so I cannot look into this. 

@dkeeney I'll try granting you the access to CI, so the bus-factor increases.

breznak commented:
It seems the CI is running - it was manually disabled, don't know why(?). 

Please fix tiny errs in Debug https://github.com/htm-community/htm.core/pull/912/checks?check_run_id=1924693127#step:4:9

breznak commented:
Sidenote: @dkeeney we should likely move to ubuntu 20.04 image (still on 18.04)

dkeeney commented:
> Sidenote: @dkeeney we should likely move to ubuntu 20.04 image (still on 18.04)

Yes I agree.  And Python 38 if it isn't already there.

ctrl-z-9000-times commented:
It appears that the Mac-OS compiler got stricter about the following issue, which is causing the CI failures:

`for (const auto &x : my_list)`
Should the reference `&` be there? It depends on whether `entries` is a reference as well?
I think that this is usually a warning, but our build system treats warnings as errors?

I fixed these issues where I could, but this issue is also in the `pybind11` dependency which I can't easily modify.

FWIW this PR passes all tests on my personal computer (Ubuntu18) in release mode. To pass the tests in debug mode I had to use `--ignore=bindings/py/tests/algorithms/spatial_pooler_test.py` because that file took too long to run, and also `TemporalMemoryBindingsTest.testPerformanceLarge` failed because it took more than a hardcoded amount of time to run.

breznak commented:
> I fixed these issues where I could, but this issue is also in the pybind11 dependency which I can't easily modify. 

thanks for the fixes in our codebase! Meanwhile, there was a large update to pybind11 which seems to fix those (warning) issues. 

Merging as-is, 1 http-lib test failed to connect, I think that's a random issue.

Finnfj commented:
Hey guys, I just ran into the error, a violation of this new check is causing.
I've looked at the code for it and I wondered if it is really that simple in every case. 
In my case I was encoding 4 network metrics with a resolution of 0.1, sparsity of 0.5 and size of 100 for each before.
I had really good anomaly detection with those parameters but now I had to tune up the size of the SDRs a lot, which changed results and performance. 

I've thought about my case if it's really that bad to have a high collision rate. That's because I am normalizing my values and they mostly only go from sth like -20 to 20 with fractions ofc. 
Are there really that many possible collisions between these values or is that a case where it would make sense to disable the check?

I'm sorry if I'm saying something stupid.

Also I am using sparsity but I've seen that the check immediately allows size above 1000 + activebits above 10?? What if I have 1000 size and 999 activeBits?

ctrl-z-9000-times commented:
> I had to tune up the size of the SDRs a lot, which changed results and performance.

I'd like to know, were the results & performance significantly better, worse, or the about same?

> Are there really that many possible collisions between these values or is that a case where it would make sense to disable the check?

Lets check

```python
size = 100
sparsity = .50
a = htm.SDR(size).randomize(sparsity)
for i in range(1000):
    b = htm.SDR(size).randomize(sparsity)
    overlap = a.getOverlap(b) / len(a.sparse)
    print(overlap)
    # assert overlap < .33
```
Further analysis of the overlaps shows that:
mean overlap: 0.50176
min overlap: 0.34
max overlap: 0.68

The threshold `33% overlap` is somewhat arbitrary.

Finnfj commented:
> I'd like to know, were the results & performance significantly better, worse, or the about same?

Went to size 1000 and had terrible results on same TM&SP settings. I have spent the day trying to tune it a little bit and have gotten to mostly similar results except it's now missing 1 of 3 DoS attacks completely while it managed to catch all 3 before.

Here's an example, one input consists of 10ms window of measurement, so the X value/100 represents seconds.
before update (which also had the faulty TM pruning leading to memory leak. Unsure if it also affects the results):
![result_Thu Apr 29 16-04-08 2021](https://user-images.githubusercontent.com/28825397/118015130-3dc6de00-b354-11eb-84be-450bfbae2510.png)

now after upping size to 1000 with some SP/TM tuning. As you can see the 3rd attack is completely missing:
![result_Wed May 12 18-16-22 2021](https://user-images.githubusercontent.com/28825397/118015219-53d49e80-b354-11eb-8857-eab5321b38b5.png)

ctrl-z-9000-times commented:
>  I have spent the day trying to tune it 

My uninformed recommendations would be to use the same number of active bits (50) and only increase the number of columns.

Finnfj commented:
> My uninformed recommendations would be to use the same number of active bits (50) and only increase the number of columns.
Thank you very much. You're not uninformed at all! It's thanks to your experience and knowledge with HTM that it looks much more usable again now. My understanding is clearly yet lacking.
